toolkit/tools/imagegen/diskutils/diskutils.go

```diff
@@ -660,3 +721,92 @@ func getPartUUID(device string) (uuid string, err error) {
+		logger.Log.Errorf("Error accessing sector size file %s", sectorFile)
+		err = ferr
+		return
+	} else if !exists {
+		err = fmt.Errorf("could not find the hw sector size file %s to obtain the sector size of the system", sectorFile)
+		return
+	}
+
+	fileContent, err := file.ReadLines(sectorFile)
+	if err != nil {
+		logger.Log.Errorf("Failed to read from %s: %s", sectorFile, err)
+		return
+	}
+
+	// sector file should only have one line, return error if not
+	if len(fileContent) != 1 {
+		err = fmt.Errorf("%s has more than one line", sectorFile)
+		return
+	}
+
+	sectorSize, err = strconv.ParseUint(fileContent[0], 10, 64)
+	return
+}
+
+func getSectorSize(diskDevPath string) (logicalSectorSize, physicalSectorSize uint64, err error) {
+	// Grab the specific disk name from /dev/xxx
+	diskName := diskDevPath[5:len(diskDevPath)]
+
+	hw_sector_size_file := fmt.Sprintf("/sys/block/%s/queue/hw_sector_size", diskName)
+	physical_sector_size_file := fmt.Sprintf("/sys/block/%s/queue/physical_block_size", diskName)
+
+	logicalSectorSize, err = getSectorSizeFromFile(hw_sector_size_file)
+	if err != nil {
+		return
+	}
+
+	physicalSectorSize, err = getSectorSizeFromFile(physical_sector_size_file)
+	return
+}
+
+func alignSectorAddress(sectorAddr, logicalSectorSize, physicalSectorSize uint64) (alignedSector uint64) {
+	// Need to make sure that starting sector of a partition is aligned based on the physical sector size of the system.
+	// For example, suppose the physical sector size is 4096. If the input start sector is 40960001, then this is misaligned,
+	// and need to be elevated to the next aligned address, which is (40960001/4096 + 1)*4096 = 4100096.
+
+	// We do need to take care of a special case, which is the first partition (normally boot partition) might be less than
+	// the physical sector size. In this case, we need to check whether the start sector is a multiple of 1 MiB.
+	if sectorAddr < physicalSectorSize {
+		if sectorAddr%(MiB/logicalSectorSize) == 0 {
+			alignedSector = sectorAddr
+		}
+	} else if (sectorAddr % physicalSectorSize) == 0 {
+		alignedSector = sectorAddr
+	} else {
+		alignedSector = (sectorAddr/physicalSectorSize + 1) * physicalSectorSize
+	}
+
+	return
+}
+
+func obtainPartitionDetail(partitionIndex int, hasExtendedPartition bool) (partType string, partitionNumber int) {
+	const (
+		indexOffsetForNormalPartitionNumber  = 1
+		indexOffsetForLogicalPartitionNumber = 2
+	)
+
+	// partitionIndex is the index of the partition in the partition array, which starts at 0.
+	// partitionNumber, however, starts at 1 (E.g. /dev/sda1), and thus partitionNumber = partitionIndex + 1.
+	// In the case of logical partitions, since an extra extended partition has to be created first in order to
+	// to create logical partitions, so the partition number will further increase by 1, which equals partitionIndex + 2.
+
+	if hasExtendedPartition && partitionIndex >= (maxPrimaryPartitionsForMBR-1) {
```

Comment:
>(maxPrimaryPartitionsForMBR-1) [](http://example.com/codeflow?start=46&length=30)

Nit: inconsistent spacing around the minus symbol.

Reply:
Same reasoning as the spacing issue mentioned above. The go formatter will throw error once fixing the spacing.